[WITH FIX] RuntimeError: Attempting to deserialize object on CUDA device 1 but torch.cuda.device_count() is 1.

**Information:**

- Chainner version: v0.18.3
- OS: Windows 11

**Description**

_I provide the fix below, this is just for documentation and for the devs if they want to fix something._

I run a laptop with an eGPU. I disabled the internal graphic card RTX 3050 and on my eGPU I have a RTX 2070 SUPER.

But for some reason, after disabling the internal gfx card, whenever I tried to load a model, I had the same error: "Model is unsupported". Though I was able to run these models before, before I disabled the internal device. It seems to me that the CUDA device 1 is selected when I only have 1 device and so it should be index 0:

`RuntimeError: Attempting to deserialize object on CUDA device 1 but torch.cuda.device_count() is 1. Please use torch.load with map_location to map your storages to an existing device.`

Even after I set the device to 0 in the settings (cf. screenshot below) the selected device was still 1. It seems chaiNNer doesn't update the configuration even though in the settings the device 0 was properly selected.

**Solution**

I had to open the `pytorch-gpu` file, from the chaiNNer folder, and change the index from 1 to 0 manually.

**Logs**

![image](https://user-images.githubusercontent.com/979397/226080513-aab40504-b492-4815-8715-cc694bd7f617.png)

```
> nvidia-smi.exe 
Sat Mar 18 10:06:50 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 528.49       Driver Version: 528.49       CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:04:00.0  On |                  N/A |
| 25%   32C    P8    24W / 215W |    475MiB /  8192MiB |     10%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

```
[2023-03-18 10:46:42.150] [info]  Backend: [16044] [ERROR] Model 003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN.pth is unsupported by chaiNNer. Please try another.
Traceback (most recent call last):
  File "{USER_DIR}\AppData\Local\chainner\app-0.18.3\resources\src\nodes\nodes\pytorch\load_model.py", line 61, in run
    state_dict = torch.load(
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 607, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 882, in _load
    result = unpickler.load()
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 857, in persistent_load
    load_tensor(data_type, size, key, _maybe_decode_ascii(location))
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 846, in load_tensor
    loaded_storages[key] = restore_location(storage, location)
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 827, in restore_location
    return default_restore_location(storage, str(map_location))
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 175, in default_restore_location
    result = fn(storage, location)
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 151, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\site-packages\torch\serialization.py", line 142, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on CUDA device '
RuntimeError: Attempting to deserialize object on CUDA device 1 but torch.cuda.device_count() is 1. Please use torch.load with map_location to map your storages to an existing device.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "{USER_DIR}\AppData\Local\chainner\app-0.18.3\resources\src\run.py", line 305, in run_individual
    raw_output, execution_time = await app.loop.run_in_executor(
  File "{USER_DIR}\AppData\Roaming\chaiNNer\python\python\lib\concurrent\futures\thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "{USER_DIR}\AppData\Local\chainner\app-0.18.3\resources\src\process.py", line 221, in wrapper
    result = supplier()
  File "{USER_DIR}\AppData\Local\chainner\app-0.18.3\resources\src\nodes\nodes\pytorch\load_model.py", line 81, in run
    raise ValueError(
ValueError: Model 003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN.pth is unsupported by chaiNNer. Please try another.
[2023-03-18 10:46:42 +0800] - (sanic.access)[INFO][127.0.0.1:52374]: POST http://127.0.0.1:8000/run/individual  200 133
```

--- a/src/renderer/components/SettingsModal.tsx
+++ b/src/renderer/components/SettingsModal.tsx
@@ -116,39 +116,59 @@ const Toggle = memo(({ title, description, isDisabled, value, onToggle }: Toggle
     );
 });
 
-interface DropdownProps extends SettingsItemProps {
+interface DropdownProps<T> extends SettingsItemProps {
     isDisabled?: boolean;
-    value: string | number;
-    options: { label: string; value: string | number }[];
-    onChange: React.ChangeEventHandler<HTMLSelectElement>;
+    value: T;
+    options: readonly { label: string; value: T }[];
+    onChange: (value: T) => void;
 }
 
-const Dropdown = memo(
-    ({ title, description, isDisabled, value, options, onChange }: DropdownProps) => {
-        return (
-            <SettingsItem
-                description={description}
-                title={title}
+// eslint-disable-next-line prefer-arrow-functions/prefer-arrow-functions, react-memo/require-memo
+function Dropdown<T>({
+    title,
+    description,
+    isDisabled,
+    value,
+    options,
+    onChange,
+}: DropdownProps<T>) {
+    const index = options.findIndex((o) => o.value === value);
+
+    useEffect(() => {
+        if (index === -1 && !isDisabled) {
+            if (options.length > 0) {
+                onChange(options[0].value);
+            }
+        }
+    }, [index, options, onChange, isDisabled]);
+
+    return (
+        <SettingsItem
+            description={description}
+            title={title}
+        >
+            <Select
+                isDisabled={isDisabled}
+                minWidth="350px"
+                value={index === -1 ? 0 : index}
+                onChange={(e) => {
+                    const optionIndex = Number(e.target.value);
+                    onChange(options[optionIndex].value);
+                }}
             >
-                <Select
-                    isDisabled={isDisabled}
-                    minWidth="350px"
-                    value={value}
-                    onChange={onChange}
-                >
-                    {options.map(({ label, value: v }) => (
-                        <option
-                            key={v}
-                            value={v}
-                        >
-                            {label}
-                        </option>
-                    ))}
-                </Select>
-            </SettingsItem>
-        );
-    }
-);
+                {options.map(({ label }, i) => (
+                    <option
+                        // eslint-disable-next-line react/no-array-index-key
+                        key={i}
+                        value={i}
+                    >
+                        {label}
+                    </option>
+                ))}
+            </Select>
+        </SettingsItem>
+    );
+}
 
 const AppearanceSettings = memo(() => {
     const { useSnapToGrid, useIsDarkMode, useAnimateChain, useViewportExportPadding } =
@@ -538,9 +558,7 @@ const PythonSettings = memo(() => {
                             }))}
                             title="PyTorch GPU"
                             value={pytorchGPU}
-                            onChange={(e) => {
-                                setPytorchGPU(Number(e.target.value));
-                            }}
+                            onChange={setPytorchGPU}
                         />
                     </VStack>
                 </TabPanel>
@@ -558,9 +576,7 @@ const PythonSettings = memo(() => {
                             }))}
                             title="NCNN GPU"
                             value={ncnnGPU}
-                            onChange={(e) => {
-                                setNcnnGPU(Number(e.target.value));
-                            }}
+                            onChange={setNcnnGPU}
                         />
                     </VStack>
                 </TabPanel>
@@ -578,18 +594,14 @@ const PythonSettings = memo(() => {
                             }))}
                             title="ONNX GPU"
                             value={onnxGPU}
-                            onChange={(e) => {
-                                setOnnxGPU(Number(e.target.value));
-                            }}
+                            onChange={setOnnxGPU}
                         />
                         <Dropdown
                             description="What provider to use for ONNX."
                             options={onnxExecutionProviders}
                             title="ONNX Execution Provider"
                             value={onnxExecutionProvider}
-                            onChange={(e) => {
-                                setOnnxExecutionProvider(String(e.target.value));
-                            }}
+                            onChange={setOnnxExecutionProvider}
                         />
                         {isUsingTensorRt && (
                             <HStack>

